src/librustc_mir/borrow_check.rs

```diff
@@ -615,29 +615,106 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
+        // for which we have a MovePath (if any); the dataflow will
+        // take care of updating its state in response to
+        // super-structure moves and reinitializations.
+        //
+        // This is scenarios 1 and 2 above.
+
+        debug!("check_if_path_is_moved part1 lvalue: {:?}", lvalue);
+        if let Some(mpi) = self.move_path_closest_to(lvalue, PrefixSet::Shallow) {
             if maybe_uninits.curr_state.contains(&mpi) {
-                // find and report move(s) that could cause this to be uninitialized
                 self.report_use_of_moved(context, desired_action, lvalue_span);
-            } else {
-                // sanity check: initialized on *some* path, right?
-                assert!(flow_state.inits.curr_state.contains(&mpi));
+                return; // don't bother finding other problems.
+            }
+            // Only query longest prefix, not further ancestors;
+            // dataflow recurs on children when parents move (to
+            // support partial (re)inits).
+            //
+            // (I.e. querying parents breaks scenario 8; but may want
+            // to do such a query based on partial-init feature-gate.)
+        }
+
+        // A move of any shallow suffix of `lvalue` also interferes
+        // with an attempt to use `lvalue`. This is scenario 3 above.
+        //
+        // (Distinct from previous loop because `lvalue` does not
+        // interfere with suffixes of its prefixes, e.g. `a.b.c` and
+        // `a.b.d`)
+
+        debug!("check_if_path_is_moved part2 lvalue: {:?}", lvalue);
+        if let Some(mpi) = self.move_path_for_lvalue(lvalue) {
+            if let Some(_) = maybe_uninits.has_any_child_of(mpi) {
+                self.report_use_of_moved(context, desired_action, lvalue_span);
+                return; // don't bother finding other problems.
+            }
+        }
+
+        // To handle scenario 4, first identify longest deref-free
+        // path underlying `lvalue`, and then ensure it is
+        // initialized. By "longest" we mean "has the most field
+        // accesses"; e.g. the longest deref-free path underlying
+        // `(*a.b.c).d` is `a.b.c`.
+
+        debug!("check_if_path_is_moved part3 lvalue: {:?}", lvalue);
+        if let Some(longest_deref_free) = max_deref_free_locally_rooted_prefix_of(lvalue) {
+            debug!("check_if_path_is_moved part3 longest_deref_free: {:?}", longest_deref_free);
+            if let Some(mpi) = self.move_path_closest_to(longest_deref_free, PrefixSet::All) {
+                if maybe_uninits.curr_state.contains(&mpi) {
+                    // find and report move(s) that could cause this to be uninitialized
+                    self.report_use_of_moved(context, desired_action, lvalue_span);
+                    return; // don't bother finding other problems.
+                }
             }
         }
     }
 
+    fn move_path_closest_to(&mut self,
```

Comment:
It would be helpful, I think, to explain what this does, and also to cover the cases in which it might return `None`. Not having read into this code recently, it's sort of not obvious to me -- for example -- why the code above can do

```rust
if let Some(mpi) = ... {
    report_error()
}
```

i.e., why is it that a `None` return value means "no error" and not "definite error"?

Reply:
good point. I will review (and perhaps add an error branch at the call-site when it returns `None`)

Reply:
(one case that kept arising, which may or may not be relevant here, are paths that are built off of a static variable; we don't put entries in the `MoveData` for those. But I've been musing that it would be better to be more explicit about the scenarios: i.e., if all the cases of `None` end up being due to a static variable at the root, we should be able to signal that explicitly in a result here, and then the client will have better info about why its sound to not issue an error.)